From patchwork Tue Apr 14 14:08:52 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Lorenzo Bianconi <lorenzo@kernel.org>
X-Patchwork-Id: 14524368
Return-Path: 
 <linux-mediatek-bounces+linux-mediatek=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 6903EF9D0D2
	for <linux-mediatek@archiver.kernel.org>;
 Tue, 14 Apr 2026 14:09:10 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Cc:To:Message-Id:
	Content-Transfer-Encoding:Content-Type:MIME-Version:Subject:Date:From:
	Reply-To:Content-ID:Content-Description:Resent-Date:Resent-From:Resent-Sender
	:Resent-To:Resent-Cc:Resent-Message-ID:In-Reply-To:References:List-Owner;
	bh=v21dmV+lYuasVw3fNqFyvXKdZD52UiXwkLFP25P9nLg=; b=aYS9+PRbve983Q576LA2bknwPv
	/HQJRnKZdSz9ZZ8xERIhXybWfk8sDvPW0qw7DtdfM1VE9314tNOkOjfLfZt0gJOf9y2nIeyP3UN/l
	TsULfN+iq7TFhABaq7BWH191yfv0SESmnUlU2kLV+4hYL/rZu49donT4LuGTnzjTi4/2bmcGqI01+
	E33ndu9m+GT0NbZNejbY7skQpLvbnCwolcjHxNHTMnC9Y8AG0X5ueFSn06nFYbr+tciOWjUWXH09I
	QSt1dj/PRQO2nqG0zHU+Qm0OirBFZZm3emdNi4PdUdW6GwW7Su5KizU6HGifizLa0Ec/lHfdHhxnw
	9N7RLP8Q==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1wCeRw-0000000HQcm-1nqK;
	Tue, 14 Apr 2026 14:09:08 +0000
Received: from tor.source.kernel.org ([172.105.4.254])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1wCeRu-0000000HQcR-381U;
	Tue, 14 Apr 2026 14:09:06 +0000
Received: from smtp.kernel.org (transwarp.subspace.kernel.org [100.75.92.58])
	by tor.source.kernel.org (Postfix) with ESMTP id 682C360018;
	Tue, 14 Apr 2026 14:09:04 +0000 (UTC)
Received: by smtp.kernel.org (Postfix) with ESMTPSA id BFD7DC19425;
	Tue, 14 Apr 2026 14:09:03 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=kernel.org;
	s=k20201202; t=1776175744;
	bh=ZxZWJI8qqfzCDhQbBvHw11oPHk45TsukiZ+I1BwbZBo=;
	h=From:Date:Subject:To:Cc:From;
	b=U1EaK8uaOIEaRDvFnhCDxrClEYZPnNEOObNFPqWT38CCV6LpArpBW27pHOhjs8xpY
	 9lXg5oT+/wGbyZpZw80tVHLODI0PsOqoGSqcR0evklDSv5shC+hy09rAu6+VZ0T4sg
	 AZtMq3d2k+CgLkNFqyUydTu8MI5vYw35ntTgQHFoH5TdQSprjx7NoN4f9gmLDEyXyC
	 HulQcpKv3bliqEowIIOs83KiyOPtgKbm5oUih1CfvkTn6DHpiSioU9psb7q6RZow0G
	 7Pgq8z5T6EtGI9NmKq6gvbAI5LgKuYQJyPFoxBSGhp2ld0E6vjQa+QGR86icTsJqCi
	 li+nRO3EMBnyA==
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Tue, 14 Apr 2026 16:08:52 +0200
Subject: [PATCH net v2] net: airoha: Wait for NPU PPE configuration to
 complete in airoha_ppe_offload_setup()
MIME-Version: 1.0
Message-Id: 
 <20260414-airoha-wait-for-npu-config-offload-setup-v2-1-5a9bf6d43aee@kernel.org>
X-B4-Tracking: v=1; b=H4sIAAAAAAAC/5WNQQ6CMBBFr0Jm7RjagFFX3sOwmNApTCQd0hbUE
 O5uww1cvv/z/9sgcRROcK82iLxKEg0F7KmCfqQwMIorDLa2l7oxFkmijoRvkoxeI4Z5wV6DlwH
 V+0nJYeK8zGhurm5aa8vIQbmbI3v5HKonBM7QlXCUlDV+D/1qjup/02rQoG+4JrLUumv7eHEMP
 J01DtDt+/4DwctuKuYAAAA=
X-Change-ID: 20260412-airoha-wait-for-npu-config-offload-setup-19d04522412d
To: Andrew Lunn <andrew+netdev@lunn.ch>,
 "David S. Miller" <davem@davemloft.net>, Eric Dumazet <edumazet@google.com>,
 Jakub Kicinski <kuba@kernel.org>, Paolo Abeni <pabeni@redhat.com>,
 Lorenzo Bianconi <lorenzo@kernel.org>
Cc: linux-arm-kernel@lists.infradead.org,
 linux-mediatek@lists.infradead.org, netdev@vger.kernel.org
X-Mailer: b4 0.14.3
X-BeenThere: linux-mediatek@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-mediatek.lists.infradead.org>
List-Unsubscribe: <http://lists.infradead.org/mailman/options/linux-mediatek>,
 <mailto:linux-mediatek-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-mediatek/>
List-Post: <mailto:linux-mediatek@lists.infradead.org>
List-Help: <mailto:linux-mediatek-request@lists.infradead.org?subject=help>
List-Subscribe: <http://lists.infradead.org/mailman/listinfo/linux-mediatek>,
 <mailto:linux-mediatek-request@lists.infradead.org?subject=subscribe>
Sender: "Linux-mediatek" <linux-mediatek-bounces@lists.infradead.org>
Errors-To: 
 linux-mediatek-bounces+linux-mediatek=archiver.kernel.org@lists.infradead.org

In order to properly enable flowtable hw offloading, poll
REG_PPE_FLOW_CFG register in airoha_ppe_offload_setup routine and
wait for NPU PPE configuration triggered by ppe_init callback to complete
before running airoha_ppe_hw_init().

Fixes: 00a7678310fe3 ("net: airoha: Introduce flowtable offload support")
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
Changes in v2:
- Check for both REG_PPE_PPE_FLOW_CFG(0) and REG_PPE_PPE_FLOW_CFG(1) to
  complete.
- Link to v1: https://lore.kernel.org/r/20260412-airoha-wait-for-npu-config-offload-setup-v1-1-f4e0aa2a5d85@kernel.org
---
 drivers/net/ethernet/airoha/airoha_ppe.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)


---
base-commit: b9d8b856689d2b968495d79fe653d87fcb8ad98c
change-id: 20260412-airoha-wait-for-npu-config-offload-setup-19d04522412d

Best regards,

diff --git a/drivers/net/ethernet/airoha/airoha_ppe.c b/drivers/net/ethernet/airoha/airoha_ppe.c
index 62cfffb4f0e5..684c8ae9576f 100644
--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -1335,6 +1335,29 @@ static struct airoha_npu *airoha_ppe_npu_get(struct airoha_eth *eth)
 	return npu;
 }
 
+static int airoha_ppe_wait_for_npu_init(struct airoha_eth *eth)
+{
+	int err;
+	u32 val;
+
+	/* PPE_FLOW_CFG default register value is 0. Since we reset FE
+	 * during the device probe we can just check the configured value
+	 * is not 0 here.
+	 */
+	err = read_poll_timeout(airoha_fe_rr, val, val, USEC_PER_MSEC,
+				100 * USEC_PER_MSEC, false, eth,
+				REG_PPE_PPE_FLOW_CFG(0));
+	if (err)
+		return err;
+
+	if (airoha_ppe_is_enabled(eth, 1))
+		err = read_poll_timeout(airoha_fe_rr, val, val, USEC_PER_MSEC,
+					100 * USEC_PER_MSEC, false, eth,
+					REG_PPE_PPE_FLOW_CFG(1));
+
+	return err;
+}
+
 static int airoha_ppe_offload_setup(struct airoha_eth *eth)
 {
 	struct airoha_npu *npu = airoha_ppe_npu_get(eth);
@@ -1348,6 +1371,11 @@ static int airoha_ppe_offload_setup(struct airoha_eth *eth)
 	if (err)
 		goto error_npu_put;
 
+	/* Wait for NPU PPE configuration to complete */
+	err = airoha_ppe_wait_for_npu_init(eth);
+	if (err)
+		goto error_npu_put;
+
 	ppe_num_stats_entries = airoha_ppe_get_total_num_stats_entries(ppe);
 	if (ppe_num_stats_entries > 0) {
 		err = npu->ops.ppe_init_stats(npu, ppe->foe_stats_dma,
